yjit/src/codegen.rs

```diff
@@ -4912,7 +4912,10 @@ fn gen_opt_new(
         // Fast path
         // call rb_class_alloc to actually allocate
         jit_prepare_non_leaf_call(jit, asm);
-        let obj = asm.ccall(rb_obj_alloc as _, vec![comptime_recv.into()]);
+
+        let allocator = unsafe { rb_get_alloc_func(comptime_recv) };
+        let allocator = allocator.unwrap_or(rb_obj_alloc);
+        let obj = asm.ccall(allocator as _, vec![comptime_recv.into()]);
```

Comment:
When an allocator function is found, it seems to skip the `if (rb_obj_class(obj) != rb_class_real(klass)) { rb_raise(rb_eTypeError, "wrong instance allocation"); }` check in `class_call_alloc_func`? Is the error unreachable?

Reply:
Wouldn't this only be a problem for T_DATA objects?  I don't think they use `rb_class_new_instance_pass_kw` as their allocator.

Reply:
I think it's only reachable from a custom allocator which is implemented incorrectly (there's no reason to ever return an object with the wrong class), so I'm not sure if it's necessary to keep.

If it is I suppose we could allowlist just the most common builtin allocators (or even just the T_OBJECT allocator).

Reply:
I honestly didn't know when it happens (it'd be nice to have a comment there) and thus asked a question. If the check is useless (on `opt_new`'s `rb_obj_alloc` call) on the interpreter too, it's fine.